File: backend/src/kms_bot/schemas/query.py
from __future__ import annotations

from pydantic import AnyUrl, Field

from kms_bot.schemas.common import StrictSchemaModel
from kms_bot.schemas.documents import ChunkRecord


class QueryRequest(StrictSchemaModel):
    query: str = Field(min_length=1)
    top_k: int = Field(ge=1, le=10)
    include_debug: bool


class SearchResultHit(ChunkRecord):
    score: float


class QuerySource(StrictSchemaModel):
    title: str = Field(min_length=1)
    url: AnyUrl
    section: str = Field(min_length=1)
    doc_id: str = Field(min_length=1)
    chunk_id: str = Field(min_length=1)


class RelatedDocument(StrictSchemaModel):
    page_id: str = Field(min_length=1)
    title: str = Field(min_length=1)
    url: AnyUrl


class QueryDebugInfo(StrictSchemaModel):
    normalized_query: str
    selected_chunks: list[SearchResultHit]


class QueryResponse(StrictSchemaModel):
    answer: str
    sources: list[QuerySource]
    related_documents: list[RelatedDocument]
    debug: QueryDebugInfo


class AnswerGeneratorInput(StrictSchemaModel):
    query: str = Field(min_length=1)
    normalized_query: str = Field(min_length=1)
    prompt_template_path: str = Field(pattern=r"^prompts/.+\.md$")
    selected_chunks: list[SearchResultHit]
    context_text: str = ""
    include_debug: bool

File: frontend/src/types/query.ts
/** Matches OpenAPI QueryRequest */
export interface QueryRequest {
  query: string;
  top_k: number;
  include_debug: boolean;
  thinking: boolean;
}

/** SSE event emitted during THINKING mode streaming */
export interface ThinkingEvent {
  stage: "planning" | "searching" | "reading" | "summarizing" | "done";
  message?: string;
  articles_found?: number;
  reading?: number;
  current?: number;
  total?: number;
  /** Present only when stage === "done" */
  data?: QueryResponse;
}

/** Matches OpenAPI QuerySource */
export interface QuerySource {
  title: string;
  url: string;
  section: string;
  doc_id: string;
  chunk_id: string;
}

/** Matches OpenAPI RelatedDocument */
export interface RelatedDocument {
  page_id: string;
  title: string;
  url: string;
}

/** Matches OpenAPI SearchResultHit */
export interface SearchResultHit {
  chunk_id: string;
  doc_id: string;
  title: string;
  section: string;
  content: string;
  url: string;
  tags: string[];
  pipeline_version: number;
  score: number;
}

/** Matches OpenAPI QueryDebugInfo */
export interface QueryDebugInfo {
  normalized_query: string;
  selected_chunks: SearchResultHit[];
}

/** Matches OpenAPI QueryResponse */
export interface QueryResponse {
  answer: string;
  sources: QuerySource[];
  related_documents: RelatedDocument[];
  debug: QueryDebugInfo;
}

/** Matches OpenAPI ErrorResponse */
export interface ErrorResponse {
  error_code: string;
  message: string;
  details?: Record<string, unknown>;
}

/** Chat message for UI state */
export interface ChatMessage {
  id: string;
  role: "user" | "assistant";
  content: string;
  timestamp: Date;
  /** Present only for assistant messages */
  response?: QueryResponse;
  /** Present when the query failed */
  error?: string;
  /** Present during THINKING mode streaming */
  thinkingStage?: ThinkingEvent;
}
